src/main/java/org/deepsymmetry/beatlink/dbserver/ConnectionManager.java

```diff
@@ -280,59 +281,76 @@ private void setPlayerDBServerPort(int player, int port) {
+            for (int tries = 0; tries < 4; ++tries) {
 
-            if (tries > 0) {
-                try {
-                    Thread.sleep(1000 * tries);  // Give the player more time to be ready, it may be booting.
-                } catch (InterruptedException e) {
-                    logger.warn("Interrupted while trying to retry dbserver port query?");
+                if (tries > 0) {
+                    try {
+                        Thread.sleep(1000 * tries);  // Give the player more time to be ready, it may be booting.
+                    } catch (InterruptedException e) {
+                        logger.info("Interrupted while trying to retry dbserver port query, must be shutting down.");
+                        return;
+                    }
                 }
-            }
 
-            Socket socket = null;
-            try {
-                InetSocketAddress address = new InetSocketAddress(announcement.getAddress(), DB_SERVER_QUERY_PORT);
-                socket = new Socket();
-                socket.connect(address, socketTimeout.get());
-                InputStream is = socket.getInputStream();
-                OutputStream os = socket.getOutputStream();
-                socket.setSoTimeout(socketTimeout.get());
-                os.write(DB_SERVER_QUERY_PACKET);
-                byte[] response = readResponseWithExpectedSize(is);
-                if (response.length == 2) {
-                    final int portReturned = (int)Util.bytesToNumber(response, 0, 2);
-                    if (portReturned == 65535) {
-                        logger.info("Player {} reported dbserver port of {}, not yet ready?", announcement.getDeviceNumber(), portReturned);
-                    } else {
-                        setPlayerDBServerPort(announcement.getDeviceNumber(), portReturned);
-                        return;  // Success!
+                Socket socket = null;
+                try {
+                    InetSocketAddress address = new InetSocketAddress(announcement.getAddress(), DB_SERVER_QUERY_PORT);
+                    socket = new Socket();
+                    socket.connect(address, socketTimeout.get());
+                    InputStream is = socket.getInputStream();
+                    OutputStream os = socket.getOutputStream();
+                    socket.setSoTimeout(socketTimeout.get());
+                    os.write(DB_SERVER_QUERY_PACKET);
+                    byte[] response = readResponseWithExpectedSize(is);
+                    if (response.length == 2) {
+                        final int portReturned = (int)Util.bytesToNumber(response, 0, 2);
+                        if (portReturned == 65535) {
+                            logger.info("Player {} reported dbserver port of {}, not yet ready?", announcement.getDeviceNumber(), portReturned);
+                        } else {
+                            if (isRunning()) {  // Bail if we were shut down before we received a response.
```

Comment:
If the thread is not sleeping when we tried to interrupt it, we can still reach this point, so we do one final check to make sure Beat Link did not shut down before we record the port number.